### PR TITLE
Improve capabilites handing and add SASL emitters

### PR DESCRIPTION
**Pull request by Half-Shot**
Depends upon #61 

This basically revamps how we do capabilities by splitting some of the code into a caps object. I'm not entirely convinced that we've gone far enough given there is a lot of logic in `irc.ts` still though.

As a side effect, I've also added support for the SASL codes in https://ircv3.net/registry

### CODE REVIEW

**Review by jaller94 — CHANGES_REQUESTED** [review]
Rejecting because of `v ?? parseInt(v)`.

File: src/irc.ts [hunk 1]
```diff
@@ -288,360 +300,633 @@ export class Client extends EventEmitter {
         if (this.opt.autoConnect === true) {
             this.connect();
         }
+    }
+
+    private onCapsList() {
```
[the remark this answers is not in the document]
Comment (Half-Shot) [reply src/irc.ts:305]: TS complains that this is undefined when I make this change?

File: src/irc.ts [hunk 2]
```diff
@@ -288,360 +300,633 @@ export class Client extends EventEmitter {
...
-                                        if (idx) {
-                                            const result = match2.shift();
-                                            if (result) {
-                                                this.prefixForMode[idx] = result;
-                                            }
-                                        }
-                                    }
+                        this.supportedState.usermodepriority = match[1];
+                        const match1 = match[1].split('');
+                        const match2 = match[2].split('');
```
[the remark this answers is not in the document]
Comment (Half-Shot) [reply src/irc.ts:418]: >Is there a better name for match1 and match2?

Renamed and added examples to it to make it easier to follow. I think in the future this whole block would be nice to split out as it's own parser function.

> Also, .split('') separates UTF16 blocks,

This is fine, I'd be surprised if there was anything other than ascii in this part. Certainly not UTF16.

File: src/irc.ts [hunk 3]
```diff
@@ -288,360 +300,633 @@ export class Client extends EventEmitter {
...
+            // No users
+            return;
+        }
+        const users = message.args[3].trim().split(/ +/);
+        users.forEach(user => {
+            // user = "@foo", "+foo", "&@foo", etc...
+            // The symbols are the prefix set.
+            const allowedSymbols = Object.keys(this.modeForPrefix).join("");
+            // Split out the prefix from the nick e.g "@&foo" => ["@&foo", "@&", "foo"]
+            const prefixRegex = new RegExp("^([" + escapeRegExp(allowedSymbols) + "]*)(.*)$");
```
[the remark this answers is not in the document]
Comment (Half-Shot) [reply src/irc.ts:617]: A lot of this code is from a long time ago, so I've not overly refactored the quotes yet. I kinda want to save that for another PR.

File: src/irc.ts [hunk 4]
```diff
@@ -288,360 +300,633 @@ export class Client extends EventEmitter {
...
+    private onInvite(message: Message) {
+        this._casemap(message, 1);
+        const from = message.nick;
+        this.emit('invite', message.args[1], from, message);
+    }
+
+    private onQuit(message: Message) {
+        if (this.opt.debug) {util.log('QUIT: ' + message.prefix + ' ' + message.args.join(' '));}
+        if (this.nick === message.nick) {
+            // TODO handle?
```
[the remark this answers is not in the document]
Comment (Half-Shot) [reply src/irc.ts:812]: Leftover from the old code, still not sure what we do here.

File: src/irc.ts [hunk 5]
```diff
@@ -288,360 +300,633 @@ export class Client extends EventEmitter {
...
-                    this.currentNick = message.args[0];
-                    this._updateMaxLineLength();
+            else {
+                let modeArg;
+                // channel modes
+                if (mode.match(/^[bkl]$/)) {
+                    modeArg = modeArgs.shift();
+                    if (!modeArg || modeArg.length === 0) {modeArg = undefined;}
                 }
+                // TODO - deal nicely with channel modes that take args
```
[the remark this answers is not in the document]
Comment (Half-Shot) [reply src/irc.ts:550]: ugh, still need to solve this one. It's a leftover from the code.

**Review by jaller94 — APPROVED** [review]
Looks good, other than the fact that you still want to work on the TODO you had left.

**Pull request merged**